cpp/include/cudf/utilities/output_writer_iterator.cuh

```diff
@@ -0,0 +1,175 @@
+template <class BinaryFunction, class Iterator>
+struct is_proxy_reference<output_writer_iterator_proxy<BinaryFunction, Iterator>>
+  : public thrust::detail::true_type {
+};
+
+}  // namespace detail
+}  // namespace thrust
+
+namespace cudf {
+/**
+ * @brief Transform output iterator with custom writer binary function which takes index and value.
+ *
+ * @code {.cpp}
+ * #include <cudf/utilities/output_writer_iterator.cuh>
+ * #include <thrust/device_vector.h>
+ * #include <thrust/iterator/counting_iterator.h>
+ * #include <thrust/iterator/transform_iterator.h>
+ *
+ * struct set_bits_field {
+ *   int* bitfield;
+ *   __device__ inline void set_bit(size_t bit_index)
+ *   {
+ *     atomicOr(&bitfield[bit_index/32], (int{1} << (bit_index % 32)));
+ *   }
+ *   __device__ inline void clear_bit(size_t bit_index)
+ *   {
+ *     atomicAnd(&bitfield[bit_index / 32], ~(int{1} << (bit_index % 32)));
+ *   }
+ *   // Index, value
+ *   __device__ void operator()(size_t i, bool x)
+ *   {
+ *     if (x)
+ *       set_bit(i);
+ *     else
+ *       clear_bit(i);
+ *   }
+ * };
+ *
+ * thrust::device_vector<int> v(1, 0x0000);
+ * auto result_begin = cudf::make_output_writer_iterator(thrust::make_counting_iterator(0),
+ *                                                 set_bits_field{v.data().get()});
+ * auto value = thrust::make_transform_iterator(thrust::make_counting_iterator(0), [] __device__
+ * (int x) { return x%2;
+ * });
+ * thrust::copy(thrust::device, value, value+32, result_begin);
+ *
+ * #include <cudf/utilities/output_writer_iterator.cuh>
+ * #include <thrust/device_vector.h>
+ * #include <thrust/iterator/counting_iterator.h>
+ * #include <thrust/iterator/transform_iterator.h>
+ *
+ * struct set_bits_field {
+ *   int* bitfield;
+ *   __device__ inline void set_bit(size_t bit_index)
+ *   {
+ *     atomicOr(&bitfield[bit_index/32], (int{1} << (bit_index % 32)));
+ *   }
+ *   __device__ inline void clear_bit(size_t bit_index)
+ *   {
+ *     atomicAnd(&bitfield[bit_index / 32], ~(int{1} << (bit_index % 32)));
+ *   }
+ *   // Index, value
+ *   __device__ void operator()(size_t i, bool x)
+ *   {
+ *     if (x)
+ *       set_bit(i);
+ *     else
+ *       clear_bit(i);
+ *   }
+ * };
+ *
+ * thrust::device_vector<int> v(1, 0x0000);
+ * auto result_begin = cudf::make_output_writer_iterator(thrust::make_counting_iterator(0),
+ *                                                 set_bits_field{v.data().get()});
+ * auto value = thrust::make_transform_iterator(thrust::make_counting_iterator(0),
+ *   [] __device__ (int x) {   return x%2; });
+ * thrust::copy(thrust::device, value, value+32, result_begin);
+ * int(v[0]); // returns 0xaaaaaaaa;
+ * @endcode
+ *
+ *
+ * @tparam BinaryFunction Binary function to be called with the Iterator value and the rhs of
+ * assignment operator.
+ * @tparam Iterator iterator type that acts as index of the output.
+ */
+template <typename BinaryFunction, typename Iterator>
+class output_writer_iterator
```

Comment:
This looks like a lot of machinery and I'm not clear about what it's purpose is.

Reply:
I created this first to test, if I can set `element` and `set_valid` at the same time using single `reduce_by_key` with `thrust::optional`. But using `thrust::optional` was always slow.
So, I reverted to use 2 `reduce_by_key`, 1 for element, another for null_mask. 

To use `reduce_by_key` with null_mask, it needs a temporary bool buffer and `valid_if`.
To avoid this, I used this `transform_output_writer_iterator`.
Anyway, this is purely to avoid allocation of temporary bool buffer, but it doesn't affect performance much. 
I will revert to using temporary bool buffer, and remove this iterator. (I thought that there may be other use cases for this iterator in cudf. Especially with null_mask.)

Reply:
This machinery is the bare minimum to use a proxy object for assignments in thrust. The question is if the proxy object is useful. If it is, then the machinery is just overhead.

A better name, imho, would be `assignment_iterator`

From what I see the use case is _overriding the assignment operation_. In an output iterator, the assignment operator is used to take the value on the rhs and assign it to the lhs. here, we intercept that assignment operation and call a binary operator `binop(lhs, rhs)` that can override the assignment operation.

This PR uses the `assignment_iterator` with a lambda that captures a null mask. The proxy intercepts the lhs `int` and a rhs `bool` and then invokes the lambda which calls `set_valid` or `set_null` on the captured mask. It is the _output_ version of `make_validity_iterator`.

Reply:
I think we should be more specific here. I would like to see a `make_validity_output_iterator` that is very focused on dealing with null masks of a particular column. With an interface like this, we can start to experiment with opportunistic coalescing of null mask assignments using cooperative groups. Also, this API makes it very obvious that we are doing _individual bit assignments_ and that coalescing them by changing the calling code could prove more performant.

```c++
auto make_validity_output_iterator(mutable_column_device_view const& destination);
```

Reply:
The implementation may well use the proxy-based iterator you've created, but there would be a very clear use case for it, and other developers will have an easy to use API for that use case.

Reply:
I also want to make sure that this proxy-based iterator functionality does not already exists in Thrust.

@allisonvacanti does thrust provide an output iterator that allows us to intercept the assignment operation and instead call a different function?

Reply:
@elstehle IIRC you've used thrust a lot... have you ever run in to an instance where this sort of assignment interception would be useful?

Reply:
WRT using cooperative groups to discovering ways we can coalesce bitmask writes, that could be a deep rabbit hole and the performanc benefits (if any) are not obvious.

Reply:
I chatted with @karthikeyann offline for a bit, and it looks like there is some existing functionality for discovering this type of concurrency! cooperative_groups (cg) provides a `labeled_partition` API which takes an integer, and creates coalesced groups on that key. You can then use cg's `reduce` function to create the new mask value. That value will need to be carefully combined with the existing bits in the null mask to make sure we don't overwrite any bits we aren't trying to assign. This looks like low hanging fruit to find out if this is more or less performant than setting individual bits atomically.

Reply:
Thrust does not offer anything like this.

Reply:
I'm highly circumspect of any iterator facade that uses `atomicOr` underneath to set individual bits as it's a huge performance footgun.

> using cooperative groups to discover ways we can coalesce bitmask writes

There's no way to use cooperative groups underneath a generic, random access output iterator to do coalesced bitmask writes. The writes to the mask can come at any time, from any thread, in any block, in any order, any number of times.

Reply:
I misunderstood. I was under the impression cooperative groups was able to detect which threads were active, and act accordingly, but it sounds like it still depends on warp synchronous execution.

Reply:
@jrhemstad I'm still confused. Why exactly is cooperative groups not an option here? I was thinking we could try `coalesced_threads()`, followed by `labeled_partition()` and `reduce()`, then use `thread_rank() == 0` to indicate which threads should write the output, and use atomics to make sure uncoordinated writes don't interfere with one another. I doubt this will be _fast_ in the majority of use cases, but I am not sure why it would not work. It would be an experiment to see if we can speed up the best-case scenario, where all/most writes are sequential.

Reply:
> . I was under the impression cooperative groups was able to detect which threads were active

They can. Determining which threads are active is not the problem. 

The problem is coordinating and detecting which threads are attempting to update the same bits in a given 4B or 8B word. For example, if any two threads in a grid want to update bits `i` and `i+1`, how do you detect that scenario without some form of communication? `coalesced_threads` doesn't help you here because whether or not the threads are coalesced doesn't tell you anything about what bits they are updating. 

Furthermore, even if the two threads are both active at the same time in the same warp (which you could never rely on) how do you detect that `t0` wants to update bit `i` and `t1` update bit `i+1` when it could just as well be `t0` updates `i` and `t1` updates `i + 1042`.

Reply:
That's where @karthikeyann 's proxy class comes in to play. It knows the index and the value, and the lambda is reasonable for using that index and value to assign the appropriate bit to a captured null mask. It appears we have enough information to _attempt_ opportunistic concurrency here. I'm not saying it will be beneficial, only that it seems possible and might be worth an experiment.

Reply:
> The problem is coordinating and detecting which threads are attempting to update the same bits in a given 4B or 8B word.
the proxy class has the index, and we can pass this to `labeled_partition`.

> For example, if any two threads in a grid want to update bits i and i+1, how do you detect that scenario without some form of communication?
For the sake of the experiment, the communication would be limited to 32 threads within a warp. Communicating across warps would require shmem, and since we're in a lambda called by thrust, I'm not sure if/how that would work.

> coalesced_threads doesn't help you here because whether or not the threads are coalesced doesn't tell you anything about what bits they are updating.
`coalesced_threads` is just to prevent UB/hang, `labeled_partition` would be responsible for determining what writes can be coalesced by using `idx/32` as the label (with some offset, if necessary).

> Furthermore, even if the two threads are both active at the same time in the same warp (which you could never rely on) how do you detect that t0 wants to update bit i and t1 update bit i+1 when it could just as well be t0 updates i and t1 updates i + 1042.
@karthikeyann 's proxy class knows this.